stackrox/stackrox
# ROX-20715: Add memory to clusterentities Store

## Description

This PR introduces additional memory about the past to the ClusterEntities store.
Additional memory is required to map network flow data arriving from Collector to the k8s deployments that might have been deleted or changed in the meantime. 

Before this PR, Sensor deleted the data about Endpoints and IPs immediately after K8s informers sent an update. However, Collector could continue sending data about past traffic for some time (with afterglow enabled, even up to 7 minutes after the connection was inactive, IP has changed, or deployment was deleted). After Sensor received such _late_ data from Collector, it was unable to map it to any known deployment which resulted in using the `External Entities` node on the graph. This representation is wrong, as the entities were not external in reality (this was a big issue for some customers who thought that their clusters are isolated from external networks).
Setting `ROX_PAST_ENDPOINTS_MEMORY_SIZE` to a positive value gives us control how long the past entities should be remembered in sensor before they are truly deleted.

Note: I plan to set `ROX_PAST_ENDPOINTS_MEMORY_SIZE=20` in a oneliner follow-up PR.

## Checklist
    - Failing `gke-qa-e2e-tests` are caused by quay issues

### N/A

## Testing Performed

### Here I tell how I validated my change

- [x] Existing NWG e2e tests in CI 
- [x] Manual tests with setting `ROX_PAST_ENDPOINTS_MEMORY_SIZE =0` and looking at the network graph.
- [x] Extensive manual tests with setting `ROX_PAST_ENDPOINTS_MEMORY_SIZE > 0` and the network graph.

Exemplary test scenario:
1. Have a constant traffic between two pods: `client->nginx`
2. Create svc with cluster IP for nginx - let the svc obtain `IP_1`
3. Delete that svc after a short time (small enough to allow the traffic to be spotted by Collector)
4. Ensure that Network Graph shows connection `client->nginx` an no `External Entities` are shown
5. Create svc with cluster IP for nginx again - let the svc obtain `IP_2`
6. Wait for the Collector Afterglow feature send a second update (up to 6 mins) for `IP_1`
7. Ensure that history in Sensor still has data about `IP_1` (possible only in debugging; requires setting `ROX_PAST_ENDPOINTS_MEMORY_SIZE` to a value of about `20`)
8. Ensure that Network Graph shows connection `client->nginx` an no `External Entities` are shown

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
